java-checks/src/main/java/org/sonar/java/checks/regex/AbstractRegexCheckTrackingMatchType.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2020 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks.regex;
+
+import java.util.List;
+import org.sonar.java.regex.RegexParseResult;
+import org.sonar.plugins.java.api.semantic.MethodMatchers;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+
+public abstract class AbstractRegexCheckTrackingMatchType extends AbstractRegexCheckTrackingMatchers {
+
+  protected enum MatchType {
+    FULL, PARTIAL, BOTH, UNKNOWN
+  }
+
+  private static final MethodMatchers PARTIAL_MATCHERS = MethodMatchers.or(
+    MethodMatchers.create()
+      .ofTypes("java.util.regex.Pattern", "java.lang.String")
+      .names("split", "splitAsStream", "asPredicate")
+      .withAnyParameters()
+      .build(),
+    MethodMatchers.create()
+      .ofTypes("java.util.regex.Matcher")
+      .names("find")
+      .withAnyParameters()
+      .build()
+  );
+
+  private static final MethodMatchers FULL_MATCHERS = MethodMatchers.create()
+      .ofTypes("java.util.regex.Pattern", "java.util.regex.Matcher", "java.lang.String")
+      .names("matches", "asMatchPredicate")
+      .withAnyParameters()
+      .build();
+
+  protected abstract void checkRegex(RegexParseResult regex, MatchType matchType);
+
+  @Override
+  protected MethodMatchers trackedMethodMatchers() {
+    return MethodMatchers.or(PARTIAL_MATCHERS, FULL_MATCHERS);
```

Comment:
this should probably be stored in a constant, because `trackedMethodMatchers()` is called way too many times, and we probably don't want to perform the `or()` all the time.

Another option could be to cache it in `AbstractRegexCheckTrackingMatchers`, and then only call the cached one.